src/ApplicationInsights.ServiceFabric.Native/Net45/DependencyTrackingModule/ServiceRemotingHeaderUtilities.cs

```diff
@@ -0,0 +1,257 @@
+        }
+
+        /// <summary>
+        /// Given the provided list of header value strings, return a list of key name/value pairs
+        /// with the provided keyName and keyValue. If the initial header value strings contains
+        /// the key name, then the original key value should be replaced with the provided key
+        /// value. If the initial header value strings don't contain the key name, then the key
+        /// name/value pair should be added to the list and returned.
+        /// </summary>
+        /// <param name="headerValues">The existing header values that the key/value pair should be added to.</param>
+        /// <param name="keyName">The name of the key to add.</param>
+        /// <param name="keyValue">The value of the key to add.</param>
+        /// <returns>The result of setting the provided key name/value pair into the provided headerValues.</returns>
+        private static string UpdateHeaderWithKeyValue(IEnumerable<string> headerValues, string keyName, string keyValue)
+        {
+            string[] newHeaderKeyValue = new[] { string.Format(CultureInfo.InvariantCulture, "{0}={1}", keyName.Trim(), keyValue.Trim()) };
+            IEnumerable<string> headerList = headerValues == null || !headerValues.Any()
+                ? newHeaderKeyValue
+                : headerValues
+                    .Where((string headerValue) =>
+                    {
+                        int equalsSignIndex = headerValue.IndexOf('=');
+                        return equalsSignIndex == -1 || TrimSubstring(headerValue, 0, equalsSignIndex) != keyName;
+                    })
+                    .Concat(newHeaderKeyValue);
+
+            return string.Join(", ", headerList.ToArray());
+        }
+
+        /// <summary>
+        /// Get the key value from the provided HttpHeader value that is set up as a comma-separated list of key value pairs. Each key value pair is formatted like (key)=(value).
+        /// </summary>
+        /// <param name="headerValues">The header values that may contain key name/value pairs.</param>
+        /// <param name="keyName">The name of the key value to find in the provided header values.</param>
+        /// <returns>The first key value, if it is found. If it is not found, then null.</returns>
+        private static string GetHeaderKeyValue(IEnumerable<string> headerValues, string keyName)
+        {
+            if (headerValues != null)
+            {
+                foreach (string keyNameValue in headerValues)
+                {
+                    string[] keyNameValueParts = keyNameValue.Trim().Split('=');
+                    if (keyNameValueParts.Length == 2 && keyNameValueParts[0].Trim() == keyName)
+                    {
+                        return keyNameValueParts[1].Trim();
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private static string TrimSubstring(string value, int startIndex, int endIndex)
+        {
```

Comment:
This function seems convoluted. Why not do just these two things to make it easier to understand?
1. value.Substring()
2. Trim() on the substring

Reply:
Well the answer is optimization. All of this gets called in the hot path. There could be hundreds or thousands of calls per minute. value.Substring and Trim both create additional strings that need to be garbage collected and string get garbage collected in gen3 (most expensive kind). We will potentially be forcing a lot of gen3 garbage collections if we are not careful (when we can be) with creating additional strings.